File: Week1/MathProblem/Java/LargestPrimeFactor.java
public class LargestPrimeFactor {
    
/*
 * Given a number N, the task is to find the largest prime factor of that number.
 * For eg, 24 = 2*2*2*3 , so output is 3
 * https://www.geeksforgeeks.org/problems/largest-prime-factor2601/1
 * T(c) - > O(sqrt(n)), S(c) -> O(1)
 */

    static long largestPrimeFactor(int N) {
        
        int largestPrime = 1;
        
        /* for non prime number we will have atleast one multiple <= sqrt(n) */
        for(int i=2;i*i<=N;i++)
        {
            /* check if i divides N evenly, do until n becomes 1 */
            while(N%i == 0)
            {
                /* update largest prime factor */
                largestPrime = N;
                /* divide N by i */
                N=N/i;
            }
        }
        
        /* if N is greater than 1, means N is itself prime, then we need to update it as largestPrime */
        if(N>1)
        {
            largestPrime = N;
        }
        
        return largestPrime;
    }

    public static void main(String[] args) {
        int N = 24;
        System.out.println("Largest prime factor for = "+ largestPrimeFactor(N));
        N = 17;
        System.out.println("Largest prime factor for = "+ largestPrimeFactor(N));
    }
}

File: Week1/MathProblem/Python/LargestPrimeFactor.py
"""
Given a number N, the task is to find the largest prime factor of that number.

https://www.geeksforgeeks.org/problems/largest-prime-factor2601/1
"""


def largestPrimeFactor(N):
    """ T(c) -> O(sqrt(n)) , S(c) -> O(1) """
    largest_prime = -1
    i = 2
    # a prime factor of number can not be more than sqrt(n)
    while i*i <= N:
        # if number is divisible by i, keep track of that i (2,3,...), divide that number with the divisor to get a odd number.
        while N % i == 0:
            largest_prime = i
            N = N//i
        
        i += 1
    
    # if number is greater than 1, that is prime factor
    if N > 1:
        largest_prime = N
        
    return largest_prime


def largestPrimeFactor2(N):
    """ T(c) -> O(sqrt(n)) , S(c) -> O(1) """
    i=2
    # for non prime number we will have atleast one factor <= sqrt(n)
    while i*i<=N:
        # even number will never be prime, discard mutiple of 2, then 3 and others
        if N%i == 0:
            N = N//i
        else:    
            i += 1
    return N


N  = 24
print(largestPrimeFactor(N))
N  = 5
print(largestPrimeFactor2(N))
